src/daft-plan/src/ops/project.rs

```diff
@@ -22,20 +25,441 @@ impl Project {
+                                .iter()
+                                .map(|t| t.unwrap().as_ref())
+                                .cloned()
+                                .collect(),
+                        }
+                        .into(),
+                    )
+                }
+            }
+        }
+    }
+}
+
+fn replace_column_with_semantic_id_aggexpr(
+    e: AggExpr,
+    subexprs_to_replace: &IndexSet<FieldID>,
+    schema: &Schema,
+) -> Transformed<AggExpr> {
+    // Constructs a new copy of this expression
+    // with all occurences of subexprs_to_replace replaced with a column selection.
+    // e.g. e := (a+b)+c, subexprs := {FieldID("(a + b)")}
+    //  -> Col("(a + b)") + c
+
+    match e {
+        AggExpr::Count(ref child, mode) => {
+            replace_column_with_semantic_id(child.clone(), subexprs_to_replace, schema).map_yes_no(
+                |transformed_child| AggExpr::Count(transformed_child, mode),
+                |_| e.clone(),
+            )
+        }
+        AggExpr::Sum(ref child) => {
+            replace_column_with_semantic_id(child.clone(), subexprs_to_replace, schema)
+                .map_yes_no(AggExpr::Sum, |_| e.clone())
+        }
+        AggExpr::Mean(ref child) => {
+            replace_column_with_semantic_id(child.clone(), subexprs_to_replace, schema)
+                .map_yes_no(AggExpr::Mean, |_| e.clone())
+        }
+        AggExpr::Min(ref child) => {
+            replace_column_with_semantic_id(child.clone(), subexprs_to_replace, schema)
+                .map_yes_no(AggExpr::Min, |_| e.clone())
+        }
+        AggExpr::Max(ref child) => {
+            replace_column_with_semantic_id(child.clone(), subexprs_to_replace, schema)
+                .map_yes_no(AggExpr::Max, |_| e.clone())
+        }
+        AggExpr::List(ref child) => {
+            replace_column_with_semantic_id(child.clone(), subexprs_to_replace, schema)
+                .map_yes_no(AggExpr::List, |_| e.clone())
+        }
+        AggExpr::Concat(ref child) => {
+            replace_column_with_semantic_id(child.clone(), subexprs_to_replace, schema)
+                .map_yes_no(AggExpr::Concat, |_| e.clone())
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use common_error::DaftResult;
+    use daft_core::{datatypes::Field, DataType};
+    use daft_dsl::{binary_op, col, lit, Operator};
+
+    use crate::{ops::Project, test::dummy_scan_node, LogicalPlan};
+
+    /// Test that nested common subexpressions are correctly split
+    /// into multiple levels of projections.
+    /// e.g.
+    /// ((a+a)+(a+a))+((a+a)+(a+a)) as x
+    /// ->
+    /// 1. aaaa+aaaa as x
+    /// 2. aa+aa as aaaa
+    /// 3: a+a as aa
+    #[test]
+    fn test_nested_subexpression() -> DaftResult<()> {
+        let source: LogicalPlan = dummy_scan_node(vec![
+            Field::new("a", DataType::Int64),
+            Field::new("b", DataType::Int64),
+        ])
+        .into();
```

Comment:
I think you've already figured this out, but `dummy_scan_node()` now returns a `LogicalPlanBuilder` rather than a `Source` node; sorry about the sneaky semantic merge conflict!

I've rewritten all of our optimization tests to directly use this builder since it makes building plans way easier/more concise.